### PR TITLE
fix: Hide open flyouts when MessageDialog is shown

### LINKED ISSUE #1929: Flyout does not light dismiss in the same way as UWP on Android.
On UWP a flyout will light dismiss when a button on the flyout is pressed, this does not happen on Android.

The flyout only dismisses when the screen outside the flyout is tapped

A sample that reproduces the issue can be found here:
https://github.com/BrianDT/UnoFlyoutIssue

Package Version(s):
Uno.UI: 2.0.512-dev.3530

Affected platform(s):

- [ ] iOS
- [x] Android
- [ ] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

Have not  been unable to determine if the issue also exists on iOS because on iOS button tap on a flyout does not execute the button command. This will be raised as a separate issue.

Visual Studio:

- [ ] 2017 (version: )
- [x] 2019 (version: )
- [ ] for Mac (version: )

Relevant plugins:

- [ ] Resharper (version: )

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #1929

## PR Type

What kind of change does this PR introduce?

- Bugfix

## What is the current behavior?

flyouts not closed when a MessageDialog is shown.

## What is the new behavior?

They are now closed, matching UWP behavior.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UWP/VisualTreeHelperHelper.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿#nullable enable
+
+using System;
+
+namespace Uno
+{
+	internal static class VisualTreeHelperHelper
```
Comment (Youssef1313) [inline src/Uno.UWP/VisualTreeHelperHelper.cs:7]: The duplication may feel weird. But two reasons:

1. Following the same naming done for `DispatcherTimerHelper` (ie, type name existing in one layer but not the otherr + "Helper"). Here, the type name in itself is "VisualTreeHelper".
2. Avoid name clashes between the two types

**Comment by MartinZikmund** [comment]
@jeromelaban could we use the extension facility for this purpose? Not sure if the setup works on non-Skia targets currently

File: src/Uno.UWP/VisualTreeHelperHelper.cs (hunk 1 above)
Comment (MartinZikmund) [reply src/Uno.UWP/VisualTreeHelperHelper.cs:7]: Please move this type to `Helpers` folder. As for the naming, maybe `VisualTreeHelperProxy` would be nicer. We could also rename the other to `DispatcherTimerProxy` - but let's see if we could even use the extension facility instead, that would be even cleaner.

Comment (Youssef1313) [reply src/Uno.UWP/VisualTreeHelperHelper.cs]: @MartinZikmund I moved it to `Helpers`. (Note that `DispatcherTimeHelper` is in `Uno` folder not `Helpers` folder)

Comment (MartinZikmund) [reply src/Uno.UWP/VisualTreeHelperHelper.cs]: While you are at it, you can move the `DispatcherTimeHelper` too, for cleanup purposes :-)

File: src/Uno.UWP/Helpers/VisualTreeHelperProxy.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Uno.Helpers
+{
+	internal static class VisualTreeHelperProxy
+	{
+		private static Action? _closeAllPopups;
+
+		public static void CloseAllPopups() => _closeAllPopups?.Invoke();
+
+		public static void SetPopupsCloser(Action closeAllPopups) => _closeAllPopups = closeAllPopups;
```
Comment (jeromelaban) [inline src/Uno.UWP/Helpers/VisualTreeHelperProxy.cs:13]:
```suggestion
		public static void SetCallAllPopupsAction(Action closeAllPopups) => _closeAllPopups = closeAllPopups;
```

**Comment by jeromelaban** [comment]
> @jeromelaban could we use the extension facility for this purpose? Not sure if the setup works on non-Skia targets currently

I don't think we need this type of registration for now, as it'll add a dictionary entry, that is not really needed. It's internal, so we can refactor as needed in the future.

**Comment by jeromelaban** [comment]
@Youssef1313  Could you rebase this PR? I also wonder if we can test that. Likely by trying find multiple flyouts if any.

**Review by jeromelaban — APPROVED** [review]

**Comment by Youssef1313** [comment]
@jeromelaban The test is failing on Wasm. I found that I was missing `CloseAllPopups` call for Wasm, but even after adding it the test is still failing, I disabled the test on Wasm for now.

**Comment by Youssef1313** [comment]
> > @jeromelaban The test is failing on Wasm. I found that I was missing `CloseAllPopups` call for Wasm, but even after adding it the test is still failing, I disabled the test on Wasm for now. Want me to open an issue tracking that?
> 
> Is the test working when executed manually ?

No, fails locally too. I have already disabled the test on Wasm.

**Pull request merged**